src/mame/drivers/smartstart.cpp

```diff
@@ -0,0 +1,310 @@
+seg_2 (R15) - PORT_D2
+seg_1 (R18) - PORT_E1
+seg_0 (R16) - PORT_E0
+
+=== Piezo buzzer: ===
+Port B, bit 1
+
+--------------------------------------------------------------------- */
+
+#include "emu.h"
+#include "cpu/avr8/avr8.h"
+#include "sound/dac.h"
+#include "rendlay.h"
+#include "screen.h"
+#include "speaker.h"
+
+class pensebem2017_state : public driver_device
+{
+public:
+	pensebem2017_state(const machine_config &mconfig, device_type type, const char *tag)
+		: driver_device(mconfig, type, tag),
+		m_maincpu(*this, "maincpu"),
+		m_dac(*this, "dac"),
+		m_out_x(*this, "%u.%u", 0U, 0U),
+		m_keyb_rows(*this, "ROW%u", 0U)
+	{
+	}
+	uint8_t m_port_b;
+	uint8_t m_port_c;
+	uint8_t m_port_d;
+	uint8_t m_port_e;
+
+	required_device<avr8_device> m_maincpu;
+	required_device<dac_bit_interface> m_dac;
+	output_finder<8, 8> m_out_x;
+	required_ioport_array<4> m_keyb_rows;
+
+	uint8_t port_r(offs_t offset);
+	void port_w(offs_t offset, uint8_t value);
+	void prg_map(address_map &map);
+	void data_map(address_map &map);
+	void io_map(address_map &map);
+	void pensebem2017(machine_config &config);
+	void update_display();
+
+protected:
+	virtual void machine_start() override;
+	virtual void machine_reset() override;
+};
+
+void pensebem2017_state::machine_start()
+{
+	// resolve handlers
+	m_out_x.resolve();
+}
+
+uint8_t pensebem2017_state::port_r(offs_t offset)
+{
+	uint8_t value;
+	int bit;
+
+	switch( offset )
+	{
+		case AVR8_IO_PORTB:
+			value = m_port_b & 0xc3;
+			for (bit=0; bit<8; bit++){
+				if (bit < 2 && !BIT(m_port_e, bit)) break;
+				if (bit >= 2 && !BIT(m_port_d, bit)) break;
+			}
+			if (BIT(m_keyb_rows[0]->read(), bit)) value |= (1 << 5);
+			if (BIT(m_keyb_rows[1]->read(), bit)) value |= (1 << 3);
+			if (BIT(m_keyb_rows[2]->read(), bit)) value |= (1 << 2);
+			if (BIT(m_keyb_rows[3]->read(), bit)) value |= (1 << 4);
+			return value;
+		default:
+			return 0xff;
+	}
+}
+
+void pensebem2017_state::port_w(offs_t offset, uint8_t data)
+{
+	switch( offset )
+	{
+		case AVR8_IO_PORTB: // buzzer + keyboard_select_rows
+			m_port_b = data;
+			m_dac->write(BIT(data, 1));
+			break;
+		case AVR8_IO_PORTC: // display
+			m_port_c = data;
+			update_display();
+			break;
+		case AVR8_IO_PORTD: // display
+			m_port_d = data;
+			break;
+		case AVR8_IO_PORTE: // display
+			m_port_e = data;
+			update_display();
+			break;
+		default:
+			break;
+	}
+}
+
+void pensebem2017_state::update_display()
+{
+	for (int digit=0; digit <= 5; digit++){
+		if (!BIT(m_port_c, 5 - digit)){
+			for (int seg=0; seg<8; seg++){
+				if (seg < 2){
+					m_out_x[7-digit][seg] = !BIT(m_port_e, seg);
+				} else {
+					m_out_x[7-digit][seg] = !BIT(m_port_d, seg);
+				}
+			}
+		}
+	}
+
+	for (int digit=6; digit <= 7; digit++){
+		if (!BIT(m_port_e, 7 - digit + 2)){
+			for (int seg=0; seg<8; seg++){
+				if (seg < 2){
+					m_out_x[7-digit][seg] = !BIT(m_port_e, seg);
+				} else {
+					m_out_x[7-digit][seg] = !BIT(m_port_d, seg);
+				}
+			}
+		}
+	}
```

Comment:
Is this for a multi-segment display?  MAME has built-in support for several common multi-segment displays.  You don’t need to output the bits individually like this.

Reply:
I've adopted PWM_DISPLAY to simplify this.